packages/salesforcedx-vscode-core/src/deploys/toolingDeploy.ts

```diff
@@ -0,0 +1,159 @@
+  public connection: Connection;
+  private apiVersion?: string;
+
+  public constructor(connection: Connection, apiVersion?: string) {
+    this.connection = connection;
+    if (apiVersion) this.apiVersion = apiVersion;
+  }
+
+  public async deploy(filePath: string) {
+    // find out what the type is from metadataregistry given the sourcepath and assign
+    this.metadataType = 'Apexclass';
+    const sourcePath = filePath.replace('-meta.xml', '');
+    const metadataPath = `${sourcePath}-meta.xml`;
+
+    const container = await this.createMetadataContainer();
+    await this.createContainerMember([sourcePath, metadataPath], container);
+    const asyncRequest = await this.createContainerAsyncRequest(container);
+    const output = await this.toolingRetrieve(asyncRequest);
+    return output;
+  }
+
+  public async createMetadataContainer(): Promise<ToolingCreateResult> {
+    const metadataContainer = (await this.connection.tooling.create(
+      METADATA_CONTAINER,
+      { Name: `VSCode_MDC_${Date.now()}` }
+    )) as ToolingCreateResult;
+
+    if (!metadataContainer.success) {
+      const deployFailed = new Error();
+      deployFailed.message = nls.localize('beta_tapi_mdcontainer_error');
+      deployFailed.name = 'MetadataContainerCreationFailed';
+      throw deployFailed;
+    }
+    return metadataContainer;
+  }
+
+  public buildMetadataField(metadataContent: string) {
+    const parser = new DOMParser();
+    const document = parser.parseFromString(metadataContent, 'text/xml');
+    const apiVersion =
+      this.apiVersion ||
+      document.getElementsByTagName('apiVersion')[0].textContent;
+    const statusNode = document.getElementsByTagName('status')[0];
+    const packageNode = document.getElementsByTagName('packageVersions')[0];
+
+    const metadataField = {
+      apiVersion,
+      ...(statusNode ? { status: statusNode.textContent } : {}),
+      ...(packageNode ? { packageVersions: packageNode.textContent } : {})
+    };
+    return metadataField;
+  }
+
+  public async createContainerMember(
+    outboundFiles: string[],
+    container: ToolingCreateResult
+  ): Promise<ToolingCreateResult> {
+    const id = container.id;
+    const metadataContent = fs.readFileSync(outboundFiles[1], 'utf8');
+    const metadataField = this.buildMetadataField(metadataContent);
+    const body = fs.readFileSync(outboundFiles[0], 'utf8');
+    const fileName = path.basename(
+      outboundFiles[0],
+      path.extname(outboundFiles[0])
+    );
+
+    const contentEntity = (await this.connection.tooling
+      .sobject(this.metadataType!)
+      .find({ Name: fileName }))[0];
+
+    const containerMemberObject = {
+      MetadataContainerId: id,
+      FullName: fileName,
+      Body: body,
+      Metadata: metadataField,
+      ...(contentEntity ? { contentEntityId: contentEntity.Id } : {})
+    };
+    const containerMember = (await this.connection.tooling.create(
+      supportedToolingTypes.get(this.metadataType!)!,
```

Comment:
I know it's v1 so not terribly critical, but we should fix this to not use the not-null assertion. It's an operator that tends to make things difficult to debug.